docs/modules/hdfs/pages/usage-guide/upgrading.adoc

```diff
@@ -0,0 +1,85 @@
+
+== Preparing HDFS
+
+HDFS must be configured to initiate the upgrade process. To do this, run the following commands in a HDFS superuser environment
+(either a client configured with a superuser account, or from inside NameNode pod):
+
+// This could be automated by the operator, but dfsadmin does not have good machine-readable output.
+// It *can* be queried over JMX, but we're not so lucky for finalization.
+
+[source,shell]
+----
+$ hdfs dfsadmin -rollingUpgrade prepare
+PREPARE rolling upgrade ...
+Preparing for upgrade. Data is being saved for rollback.
+Run "dfsadmin -rollingUpgrade query" to check the status
+for proceeding with rolling upgrade
+  Block Pool ID: BP-841432641-10.244.0.29-1722612757853
+     Start Time: Fri Aug 02 15:49:12 GMT 2024 (=1722613752341)
+  Finalize Time: <NOT FINALIZED>
+
+$ # Then run query until the HDFS is ready to proceed
+$ hdfs dfsadmin -rollingUpgrade query
+QUERY rolling upgrade ...
+Preparing for upgrade. Data is being saved for rollback.
+Run "dfsadmin -rollingUpgrade query" to check the status
+for proceeding with rolling upgrade
+  Block Pool ID: BP-841432641-10.244.0.29-1722612757853
+     Start Time: Fri Aug 02 15:49:12 GMT 2024 (=1722613752341)
+  Finalize Time: <NOT FINALIZED>
+
+$ # It should look like this once ready
+$ hdfs dfsadmin -rollingUpgrade query
+QUERY rolling upgrade ...
+Proceed with rolling upgrade:
+  Block Pool ID: BP-841432641-10.244.0.29-1722612757853
+     Start Time: Fri Aug 02 15:49:12 GMT 2024 (=1722613752341)
+  Finalize Time: <NOT FINALIZED>
+----
+
+== Starting the upgrade
+
+Once ready, the HdfsCluster can be updated with the new product version:
+
+[source,shell]
+----
+$ kubectl patch hdfs/simple-hdfs --patch '{"spec": {"image": {"productVersion": "3.4.0"}}}' --type=merge
+hdfscluster.hdfs.stackable.tech/simple-hdfs patched
+----
+
+Then wait until all pods are ready and running the new HDFS version.
+
+NOTE: Services will be upgraded in order: JournalNodes, then NameNodes, then DataNodes.
+
+== Finalizing the upgrade
+
+Once all HDFS pods are running the new version, the HDFS upgrade can be finalized (from the HDFS superuser environment):
+
+[source,shell]
+----
+$ hdfs dfsadmin -rollingUpgrade finalize
+FINALIZE rolling upgrade ...
+Rolling upgrade is finalized.
+  Block Pool ID: BP-841432641-10.244.0.29-1722612757853
+     Start Time: Fri Aug 02 15:49:12 GMT 2024 (=1722613752341)
+  Finalize Time: Fri Aug 02 15:58:39 GMT 2024 (=1722614319854)
+----
+
+// We can't safely automate this, because finalize is asynchronous and doesn't tell us whether all NameNodes have even received the request to finalize.
+
+WARNING: Please ensure that all NameNodes are running and available before proceeding. NameNodes that have not finalized yet will crash on launch when taken out of upgrade mode.
+
+Finally, the operator and cluster should be taken out of upgrade mode, by marking the HdfsCluster as upgraded to the new version:
```

Comment:
Unclear terminology here.. we essentially have three different "upgrade modes" that we need to keep track of:

1. HDFS cluster upgrade mode (managed by `dfsadmin`)
2. NameNode upgrade mode (allows the NameNode to launch despite not matching the version of its local fsimage)
3. HdfsCluster stacklet upgrade mode (*currently* manages 2 but not 1, ideally eventually it would manage both)

Reply:
Which terminology is unclear?

I've read through the whole doc from start to finish (hence new comments :sweat_smile:). The instructions seem clear to me.

Reply:
Mostly the two different upgrade modes, and what controls which.

Reply:
I guess they are more _phases_ than _modes_. All of them required, some parts manual, some parts done by the operator.

I think that is quite clear in the instructions, but I guess if it needs more clarity, the text at the top could be reworked?

`HDFS currently requires a manual process to upgrade.`

Could be

`HDFS currently requires manual steps to both begin and finalize process. The HDFS Operator will take care of bringing up the NameNodes in a filesystem compatibility mode` (or whatever the correct way of saying that is).

Reply:
> I guess they are more phases than modes.

Sort of. Phases from the POV of the overall process, but modes from the POV of each daemon.

> HDFS currently requires manual steps to both begin and finalize process. The HDFS Operator will take care of bringing up the NameNodes in a filesystem compatibility mode (or whatever the correct way of saying that is).

Okay, I quite like compatibility mode for the NameNodes.

Reply:
f0df2b7f49e5c9a32f9c701c6d1c342ec2127f0e